.github/workflows/docker.yml

```diff
@@ -0,0 +1,81 @@
+name: Create and publish a Docker images
```

Comment:
Overall, I see a heated discussion about technical approach, but I would love to have the discussion about approach to have this in place or not, and put effort into development, but then also maintenance, only afterwards.

Overall, we shall see the needed usecases and comapre with what is available today.

> For end-users this should make the use much easier, as no installation is needed.

somehow, end user needs to "install" Fixer. Doesn't matter if it's via Composer, via Docker, via anything else - some "action" is needed. only tool behind that action is different.

- challenge: deps conflicts
  - solution in place: instal phar file (manually, via composer `php-cs-fixer/shim`, via phive)
- but I really need dockerised env
  - solution in place: [phpqa](https://hub.docker.com/r/jakzal/phpqa/) docker image with different PHP version to use, providing newest Fixer
- or want to use in CI
  - solution: beside guidance how to do it with really tweaked setup, we also have out of the box integratinos as 3rd party implementations - eg https://github.com/OskarStark/php-cs-fixer-ga , also with newest Fixer

Overall, I see very limited benefit of "full control" for "centralised" solution replacing already existing options.
And those existing solution came to live long ago, with many learnings and optimizations on the go, lot of capacity already put there to make it shine. I doubt we have as much capacity to re-do it from scratch, and even if - I would rather put it to other remaining topics we have.

Unfortunately, this is not a discussion i have capacity to conduct in December, but feel free to pre-discuss it already.

Reply:
@keradus this PR provides feature that none of stuff mentioned by you does (in terms of combined set of advantages this approach has).

@jakzal's PHPQA image provides only one specific version of fixer per PHP version and you don't have any control which version you want to use:

![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/9c35aab7-a9ef-44d5-8a8d-4fbc4bf48738)

It also requires downloading image with total size of 1.0 GB, which is massive overkill to run Fixer. From `dive`:

```
Image name: jakzal/phpqa
Total Image size: 1.0 GB
Potential wasted space: 19 MB
Image efficiency score: 98 %
```

Even if you consider sizes presented [on the hub](https://hub.docker.com/r/jakzal/phpqa/tags) (~300MB), then fine-crafted Fixer images are 3x smaller:

```
Image name: ghcr.io/elbformat/php-cs-fixer:1.2.3-php8.3
Total Image size: 108 MB
Potential wasted space: 1.1 MB
Image efficiency score: 99 %
```

On the other hand mentioned GitHub Actions are limited to Github, obviously, but there is Gitlab CI and other CI systems that could utilise these official images. @OskarStark's image also does not have ability to choose Fixer's version, while @shivammathur's solution, even though it's great, is not a Docker replacement (you can't use it locally).

So, let's sum it up:
- GitHub's container registry is free for public repos, so there's no cost for storing images
- we can have fine-tuned, simple and lightweight images that can be used in any environment (local, CI), having PHP version / Fixer version matrix to choose from
- maintenance is really low - the built solution should just work and we'll see it in Actions (on pushes to `master` and on new tags)

> Overall, I see very limited benefit of "full control" for "centralised" solution replacing already existing options.
And those existing solutions came to live long ago, with many learnings and optimizations on the go, lot of capacity already put there to make it shine. I doubt we have as much capacity to re-do it from scratch, and even if - I would rather put it to other remaining topics we have.

Some of these tools were built only because there wasn't official image, and with the introduction of it can be considered as obsolete _OR_ can utilise our images in order to build their images 😉. For example @OskarStark's GH action could benefit from this feature and could provide config option to choose PHP version and Fixer version, with a fallback to latest release on latest supported PHP.

There's not much optimisations that can be done here (maybe putting pre-built PHAR file in the images based on tags instead of copying the code and vendors). It's just a build of PHP runtime with a Fixer, that can be used on any platform with a Docker.

I see a great value here and I 100% support this PR 👍.

Reply:
As a maintainer of the phpqa docker imager my recommendation is to use composer to install php tools. My image was created with a specific edge case in mind - to run inspections on legacy projects in order to efficiently give feedback on clients’ projects. In 99.999% of regular projects, why use anything else but composer?

Reply:
@jakzal mostly because of potential Composer conflicts. Yeah, it can be worked around with the installation of shim package, but still having a Docker image opens new way for running Fixer, which IMHO is great. Even for legacy projects (that may or may not use Composer) it's a value, mostly because of PHP version matrix available - Fixer most probably [shouldn't be run on PHP 8.3 when you're fixing old codebase](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/7526#issuecomment-1846950421), current production runtime version should be used instead. So even if Composer installation (regular or shim) is valid most of the time, having ready-to-use Docker runtimes is a great addition 🙂.

Reply:
I wouldn't have create this pull request if there wasn't a useful scenario there for me. We use gitlab-ci to build our project with "docker build". There are no dev-dependecies inside this image, as we want to test it as it is shipped. So the only way would be to download and install php-cs-fixer on each pipeline run inside our app image or on a blank php image if we would use the composer/phar way.

With the two lines of gitlab-ci config provided in the examples I have fast (no installation), lightweight (see image sizes) and flexible (easily change versions or php versions) solution, that will add no overhead (like transfering composer caches) to our CI.

I already have this docker solution in place, with the downside, that I need to build and update the images by myself manually for each project. So automation of the docker creation process at the source of where releases are created makes much more sense and will add no extra work to you imho.

Reply:
thanks for good discussion folks. I wasn't able to go by each detail here, but @Wirone gave me nice overview and we aligned on first scope here, potentially to extend in future.
thanks for working on this ;)

no further concerns from my side ;)
[probably like jakzal, I would still run composer or use shim myself, but yeah, we already support multiple options of obtaining the tool ;) ]